File: code/http/http_client.py
import socket


def main():
    host = 'httpbin.org'
    port = 80

    client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    print(f'正在连接 {host}:{port}...')
    client.connect((host, port))
    print('--- TCP 连接已建立 ---')

    request = (
        'GET /get HTTP/1.1\r\n'
        f'Host: {host}\r\n'
        'User-Agent: hello-network/1.0\r\n'
        'Accept: */*\r\n'
        'Connection: close\r\n'
        '\r\n'
    )

    print(f'\n发送请求:\n{request}')
    client.sendall(request.encode())

    response = b''
    while True:
        chunk = client.recv(4096)
        if not chunk:
            break
        response += chunk

    print(f'收到响应 ({len(response)} 字节):')
    print(response.decode(errors='replace'))

    client.close()
    print('\n--- 连接已关闭 ---')


if __name__ == '__main__':
    main()

File: code/http/HttpClient.java
import java.io.*;
import java.net.*;

/**
 * HTTP 客户端示例
 * 展示通过 TCP Socket 发送 HTTP 请求和接收响应
 *
 * 运行方式:
 *   javac HttpClient.java && java HttpClient
 */
public class HttpClient {

    public static void main(String[] args) throws IOException {
        String host = "httpbin.org";
        int port = 80;

        System.out.println("正在连接 " + host + ":" + port + "...");
        try (Socket socket = new Socket(host, port);
             PrintWriter out = new PrintWriter(
                     socket.getOutputStream(), true);
             BufferedReader in = new BufferedReader(
                     new InputStreamReader(socket.getInputStream()))) {

            System.out.println("--- TCP 连接已建立 ---\n");

            out.print("GET /get HTTP/1.1\r\n" +
                     "Host: " + host + "\r\n" +
                     "User-Agent: hello-network/1.0\r\n" +
                     "Accept: */*\r\n" +
                     "Connection: close\r\n" +
                     "\r\n");
            out.flush();

            System.out.println("响应:");
            String line;
            while ((line = in.readLine()) != null) {
                System.out.println(line);
            }
        }
        System.out.println("\n--- 连接已关闭 ---");
    }
}
